### PR TITLE
[Design] Add `ModelStateEntry` property to `ModelBindingContext`

### PR DESCRIPTION
**Pull request by dougbu**
- #4362
- WIP on potential perf improvements

### CODE REVIEW

**Comment by dougbu** [comment]
A few points keep this in design
- unclear the overall change is sufficiently beneficial
- missing doc comments
- missing unit tests of the new methods / properties
- no tests demonstrating that alternate subclasses of `ModelBindingContext` work correctly

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs [hunk 1]
```diff
@@ -68,15 +68,38 @@ public abstract class ModelBindingContext
         public abstract string ModelName { get; set; }
 
         /// <summary>
-        /// Gets or sets the <see cref="ModelStateDictionary"/> used to capture <see cref="ModelState"/> values
-        /// for properties in the object graph of the model when binding.
+        /// Gets or sets the <see cref="ModelStateDictionary"/> used to capture
+        /// <see cref="ModelBinding.ModelStateEntry"/> values for properties in the object graph of the model when
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs]: Old comment referenced the `ModelState` property in this documentation of that property.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs [hunk 2]
```diff
@@ -904,6 +914,224 @@ public ModelStateNode GetOrAddNode(StringSegment subKey)
...
+
+                if (key == null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                return GetValidity(this) ?? ModelValidationState.Unvalidated;
+            }
+
+            public override void SetModelValue(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs]: Probably some duplicate code in the `ModelStateDictionary` methods can be eliminated using these new methods.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs [hunk 3]
```diff
@@ -114,30 +119,49 @@ public BodyModelBinder(IList<IInputFormatter> formatters, IHttpRequestStreamRead
...
+                if (bindingContext.IsTopLevelObject &&
+                    !string.Equals(bindingContext.ModelName, modelBindingKey, StringComparison.Ordinal))
+                {
+                    // Cannot use ModelStateEntry tracked in ModelBindingContext due to "special logic" above.
+                    modelState.TryAddModelError(modelBindingKey, exception, metadata);
+                }
+                else
+                {
+                    bindingContext.ModelStateEntry.TryAddModelError(modelState, modelBindingKey, exception, metadata);
+                }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs]: This is about the ugliest change need to use the new API. Happens once more in this class (or as @Eilon overwhelmingly prefers `this` `class`) and again in `FormFileModelBinder` -- the two most special-casey binders.

**Comment by dougbu** [comment]
##### BigModelBinding
Exercises [`HomeController.Index()`](https://github.com/aspnet/Performance/blob/dev/testapp/BigModelBinding/Controllers/HomeController.cs#L8-L21) view using [`./loadtest.ps1`](https://github.com/aspnet/Performance/blob/dev/testapp/BigModelBinding/loadtest.ps1). Binds an instance of the [`Model`](https://github.com/aspnet/Performance/blob/dev/testapp/BigModelBinding/Models.cs#L140-L152) using form data from [`./postdata.txt`](https://github.com/aspnet/Performance/blob/dev/testapp/BigModelBinding/postdata.txt).

###### Cost
Changes eat a bit more memory because
- `DefaultModelBindingContext+State` has two more public fields; therefore &hellip;
 - `DefaultModelBindingContext` is larger because its `_state` variable is larger
 - the array backing the `Stack<State>` inside `DefaultModelBindingContext` is also larger
- code allocates `ModelStateNode`s occasionally when it will not actually be used
 - while `ModelStateNode` remains the same size, we allocate more of them
  - in `BigModelBinding` scenario, allocate 11 (about 10%) more than usual number
 - also have larger arrays backing `List<ModelStateNode>` i.e. entries for the unused `Children`
 - this _may_ indicate a slight bug in my code or unintended cases where our binders recurse unnecessarily

###### `ModelBindingContext`-related Before
modelbindingcontext - baseline

###### `ModelBindingContext`-related After
modelbindingcontext - tree walk 4362

###### `ModelState`-related Before
Below repeats a couple of entries shown in  "`ModelBindingContext`-related Before" pic.
modelstate - baseline

###### `ModelState`-related After
Below repeats a couple of entries shown in  "`ModelBindingContext`-related After" pic.
modelstate - tree walk 4362

**Comment by dougbu** [comment]
FYI extra `ModelStateNode` allocations all relate to the complex checks done in `ComplexTypeModelBinder`. That class may need to confirm if any property bound from a value provider has associated data. It must enter a new binding scope for each such property and that allocates a `ModelStateNode` for the POCO instance even when no data is available for any property. Example cases include the POCO element of a collection one past the last submitted index and `null` POCO properties of POCO objects.

**Comment by dougbu** [comment]
Not seeing consistent percentages in dotTrace results before and after using the same `BigModelBinding` scenario. But the numbers did look lower for model binding both times through. Not sure whether there's anything conclusive here but could run RPS comparisons on Azure&hellip;
##### Run 1
Running `loadtest -k -n 100000 -c 16 -T application/x-www-form-urlencoded -p postdata.txt -H "Accept: application/json; q=0.9, application/xml; q=0.6" -H "Accept-Charset: utf-8" http://127.0.0.1:5000` and folding the "All Calls" data together shows:
###### Before
bigmodelbinding trace - baseline i
###### After
bigmodelbinding trace - tree walk 4362 i
##### Run 2
Doing the same thing again but for 200000 iterations (`-n 200000`) shows:
###### Before
bigmodelbinding trace - baseline ii
###### After
bigmodelbinding trace - tree walk 4362 ii

**Comment by dougbu** [comment]
@rynowak what are your thoughts? if you're not sure, what additional data would you like?

@Eilon the answers to the questions in #4362 are:

1. Speed benefit is inconclusive at this point but definitely increases memory use.
2. No breaking change is required though the new `ModelStateEntry` methods are slightly ugly in service of back-compatibility.

**Comment by rynowak** [comment]
If this increases allocations, and has no effect on straight line performance then it's not something we want.

**Comment by rynowak** [comment]
My reading of the sampling data is that `SimpleTypeModelBinder` got faster and `ComplexTypeModelBinder` got slower by almost the same amount. If you really know to know what the absolute difference is you'd have to look at throughput in a controller environment.

Looking over the code, I think this change has blown its complexity budget and doesn't show enough benefit. /cc @Eilon

**Comment by dougbu** [comment]
Glad I was reading the dotTrace info correctly 😺 Not taking this forward.

**Pull request closed without merging**